python/mxnet/module/module.py

```diff
@@ -539,6 +543,29 @@ def forward(self, data_batch, is_train=None):
             Default is ``None``, which means ``is_train`` takes the value of ``self.for_training``.
         """
         assert self.binded and self.params_initialized
+
+        # If start to inference, force rebind module.
+        if self._label_shapes and not data_batch.label:
+            raise RuntimeError("Currently module is bind with label while new data batch "
+                               "has no label. Are you trying to inference? Rebind module "
+                               "with 'force_rebind=True' and 'for_training=False'.")
+
+        for idx in range(len(self._data_shapes)):
+            curr_data_shape = self._data_shapes[idx].shape
+            new_data_shape = data_batch.data[idx].shape
+            #Reshape if data shape changes.
+            if curr_data_shape != new_data_shape:
+                new_shape = data_batch.provide_data \
+                    if hasattr(data_batch, "provide_data") and data_batch.provide_data else \
+                    [(name, data.shape) for name, data in zip(self._data_names, data_batch.data)]
```

Comment:
also need to handle dtype and layout. Layout can be found from corresponding self._data_shapes